File: rcs_calc.py
''' 
This python script calculates the RCS reduction of a metasurface using both matlab and numpy
Author: Achintha Ihalage
Date:	06/04/2020
'''

# import matlab.engine
import numpy as np
import copy

# state = np.random.randint(4, size=(6,6)).astype(float) 
# state = np.ones((8,8))
# state = np.random.randint(2,size=(8,8)).astype(float)*3
# def build_chessboard(w, h) :
#       re = np.r_[ w*[0,1] ]              # even-numbered rows
#       ro = np.r_[ w*[1,0] ]              # odd-numbered rows
#       return np.row_stack(h*(re, ro))
# state = build_chessboard(8,8)*3.0

# metasurface proposed in the paper
# state = np.array([[1, 0, 1, 0, 1, 1, 0, 0], 
# 					[0, 0, 1, 0, 0, 0, 1, 0], 
# 					[1, 0, 0, 0, 1, 0, 1, 0], 
# 					[1, 0, 0, 1, 0, 0, 1, 1], 
# 					[0, 0, 1, 1, 1, 1, 1, 1], 
# 					[0, 0, 0, 0, 1, 0, 0, 0], 
# 					[1, 1, 0, 1, 0, 1, 0, 1], 
# 					[0, 1, 1, 0, 0, 1, 0, 0]])
# state = state*3.0
class RCS():

	def __init__(self, metasurface_width, metasurface_height):
		###################### all matlab.engine calls had to be commented out due to basic_string::_M_construct null not valid ERROR ####################
		# self.eng = matlab.engine.start_matlab()
		f = 10e9	# 10 GHz frequency
		# self.m = matlab.double([metasurface_width])
		# self.n = matlab.double([metasurface_height])
		c = 3e+8
		# self.lmd = matlab.double([c/f])
		# self.d = self.lmd 	# d=lmd
		# now define the same parameters to be used with numpy
		self.m_np = metasurface_width
		self.n_np = metasurface_height
		self.lmd_np = c/f
		self.d_np = self.lmd_np
		self.k_np = 2*np.pi/self.lmd_np # wave vector

		# self.metasurface_np = state 	# for numpy
		# self.metasurface = matlab.double(state.tolist())	# for matlab

	# def get_RCS_matlab(self, state):
	# 	metasurface_matlab = matlab.double(state.tolist())	# for matlab
	# 	try:
	# 		rcs = self.eng.rcs_calc(metasurface_matlab,self.m,self.n,self.lmd,self.d,nargout=1)
	# 		return rcs
	# 	except Exception as e:
	# 		raise e

	def get_RCS_numpy(self, state):
		Ntheta=180.0
		Nphi=180.0
		dtheta=np.pi/(2*Ntheta)
		dphi=2*np.pi/Nphi
		theta=np.linspace(0,np.pi/2,Ntheta)
		phi=np.linspace(0,2*np.pi,Nphi)
		[THETA, PHI]=np.meshgrid(theta,phi)


		# ######## get the metasurface based on the state returned by python code
		# # 1 ---> 0
		# # 2 ---> pi/4
		# # 3 ---> pi/2
		# # 4 ---> pi
		reflect_phi = copy.deepcopy(state)
		reflect_phi[(reflect_phi!=1) & (reflect_phi!=2) & (reflect_phi!=3)]=0	# assign 0 response to all other places
		reflect_phi[reflect_phi==1]=np.pi/4
		reflect_phi[reflect_phi==2]=np.pi/2
		# reflect_phi[reflect_phi==3]=3*np.pi/4
		reflect_phi[reflect_phi==3]=np.pi

		F=0
		for q in range(self.n_np):
			for p in range(self.m_np):
				F2 = np.exp((-1j)*(reflect_phi[q,p]+np.multiply(self.k_np*self.d_np*np.sin(THETA),((q-0.5)*np.cos(PHI)+(p-0.5)*np.sin(PHI)))))
				F=F+F2

		F3=np.multiply(np.abs(np.multiply(F,F)),np.sin(THETA))
		F4=np.multiply(F,F)

		Ravg=np.sum(np.sum(F3*dtheta)*dphi)
		F5=F4/Ravg

		# maximum directivity
		Rm=np.max(np.abs(F5))
		# RCS reduction
		rcs = (self.lmd_np**2)*Rm/(4*np.pi*self.m_np*self.n_np*self.d_np**2)
		return rcs

# rcs = RCS(8,8)
# # rcs_matlab = rcs.get_RCS_matlab(state)
# rcs_numpy = rcs.get_RCS_numpy(state)

# # print (rcs_matlab)
# print (rcs_numpy)
# print (state)

File: rcs_calc.m
function [RCS] = rcs_calc(metasurface,m,n,lmd,d)

%lmd=double(lmd);
%d=double(d);
k=2*(pi)/lmd;

Ntheta=180; Nphi=180;
%syms theta phi
dtheta=pi/(2*Ntheta); dphi=2*pi/Nphi;
theta=linspace(0,pi/2,Ntheta);
phi=linspace(0,2*pi,Nphi);
%phi=0;
[THETA, PHI]=meshgrid(theta,phi);



% uncomment this section for a random metasurface

%elements=[0,1]; % 2 elements corresponding to 0, pi phase responses
%idx=randi(length(elements), m,n); % indexes to create NxN random metasurface with 2 elements
%metasurface_actual = pi.*metasurface;

%%%%%%%% get the metasurface based on the state returned by python code
% 1 ---> 0
% 2 ---> pi/4
% 3 ---> pi/2
% 4 ---> pi
reflect_phi = metasurface;
reflect_phi(reflect_phi==1)=0;
reflect_phi(reflect_phi==2)=pi/4;
reflect_phi(reflect_phi==3)=pi/2;
reflect_phi(reflect_phi==4)=pi;

%reflect_phi = pi.*metasurface; % resulting metasurface



%sequence = [0,1,0,1,0,1,0,1];
%metasurface = [1 0 1 0 1 %0 0 0; 0 0 1 0 0 0 1 0; 1 0 0 0 1 0 1 0; 1 0 0 1 0 0 1 1; 0 0 1 1 0 1 1 1; 0 0 0 0 1 0 0 0; 1 1 0 1 0 1 0 1; 0 1 1 0 0 1 0 0];
%metasurface = [1 0 1 0; 0 1 1 0; 1 0 0 1; 0 1 0 1];
%reflect_phi = pi.*metasurface;


F=0;
for q=1:n
   for p=1:m
   % if( mod(q,2)+ mod(l,2)==1 )
    % u=(unidrnd(2,1)-1)*180/180*pi;
  %  w=x(q,1);v=y(q,1);
  %   u=z5(q,1);
   %u=zp((p-1)*N+q,1);
   % if(mod(q,2)==0 )
    %    F2=exp(-(1j)*(k*d*sin(THETA).*((q-0.5)*cos(PHI)+(l-0.5)*sin(PHI))));
    % else
    
       F2=exp(-(1j)*(reflect_phi(q,p)+k*d*sin(THETA).*((q-0.5)*cos(PHI)+(p-0.5)*sin(PHI))));
       %F2=exp(-(1j)*((u)+k*d*sin(THETA).*((p-0.5)*cos(PHI)+(q-0.5)*sin(PHI))));    
  
       %     F2=exp(-(1j)*((u)+(k)*sin(THETA).*(w*cos(PHI)+v*sin(PHI))));
    % end
    F=F+F2;
  end
end

   
 % F=cos(THETA).*F;
  F3=abs(F.*F).*sin(THETA);
  F4=F.*F;
  
    %R=F(THETA,PHI);
    Ravg=sum(sum(F3*dtheta)*dphi);
    F5=F4/Ravg;
    
    % maximum directivity
    Rm=max(abs(F5(:)));
    
    RCS = (lmd^2)*Rm/(4*pi*m*n*d^2);
    
    
%F4=abs(F3).*abs(F3);
%Rm=max(abs(F(:)));
%Rm=max(abs(F4(:)));
%F=20*log10(abs(F));
   
%Rm1=max(abs(F4(:)/Ravg));

%F=F-max(F);
 %plot(theta,F)
 %axis([0 pi/2 -40 0])
%Rm1=abs(Ravg);

%% uncomment for plotting
%[x,y,z]=sph2cart(PHI,pi/2-THETA,abs(F5));
%figure(1);
%rotate3d on;
%mesh(x,y,z);

end